### PR TITLE
Issue#21

### LINKED ISSUE #10: nwtopt master script
currently the user has to run a series of command to get NWTOPT up and running. This is not the intended functionality of NWTOPT and the master NWTOPT.py script needs to be fixed and reworked.

### LINKED ISSUE #19: change print() calls to logging() calls
Simple fix, more professional, should be done soon

### LINKED ISSUE #21: OOP Switch
Currently the codebase is messy, confusing, uses globals, and overall a little gross. Planning on creating an object orientated approach in the solution to this issue.

### PR DESCRIPTION
**Pull request by maxwnewcomer**
This will close #21, close #19, and close #10

### CODE REVIEW

File: NWTOPT.py [hunk 1]
```diff
@@ -20,110 +20,118 @@
...
+        self.processes = defaultdict(OPTSubprocess)
+        self.logger = self.init_logger()
+        
+        self.event_loop = asyncio.new_event_loop()
+        
+        self.log(f'Working out of {self.cwd}', 0)
+
+    def init_logger(self):
+        logger = logging.getLogger('NWTOPT')
+        logger.setLevel(logging.DEBUG)
```
Comment (mnfienen) [inline NWTOPT.py:58]: I will look as this progresses, but first look suggests maybe this `logger` object could get assigned to the `NWTOPT` object through `self`?

File: config/HParams.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            hp.uniform('dbdkappa', .00001, .0001),
+            hp.uniform('dbdgamma', 0., .0001),
+            hp.uniform('momfact', 0., .1),
+            hp.choice('backflag', [0, 1]),
+            hp.quniform('maxbackiter', 10, 50, 1),
+            hp.uniform('backtol', 1., 2.),
+            hp.uniform('backreduce', .00001, 1.),
+        ]
+        self.filetype = 'nwt'
+    def HParam2File(self, inputHp, filepath):
```
Comment (mnfienen) [inline config/HParams.py:45]: could make `filepath` and `inputHp` also attributes of the `MF5HParams` object to avoid having to pass them explicitly

File: nwtopt.sh [hunk 3]
```diff
@@ -21,12 +21,12 @@ conda-unpack
 python3 --version
 echo ...Changed Python Environment...
 echo
-mv NWT_SUBMIT/NWTOPT_FILES/objective.py nwtenv/bin/objective.py
```
Comment (mnfienen) [inline nwtopt.sh:24]: good change :)

File: objects/Condor.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        super().__init__('Condor', id, logger)
+        self.ip = ip
+        self.port = port
+        self.poll_interval = poll_interval
+        self.workers = workers
+        self.timeout = timeout
+        self.cwd = cwd
+    
+    async def init_condor(self):
+        self.modify_timeout()
```
Comment (mnfienen) [inline objects/Condor.py:17]: I can't really speak to the `async` stuff since I haven't used it before, but looks like a good approach!

File: nwtopt.sub [hunk 5]
```diff
@@ -13,10 +13,10 @@ error                   = logs/errors/errors_$(CLUSTER)_$(PROCESS).txt
 stream_output           = True
 stream_error            = True
 should_transfer_files   = Yes
-arguments               = <ip>:<port>/db 400
+arguments               = 130.11.177.99:27017/db 240
```
Comment (mnfienen) [inline nwtopt.sub:16]: A couple things 0) not good to have an internal IP address in a public repo and 1) don't you want this the way you had it so that users specify IP and PORT per submit?

File: objects/Condor.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        self.modify_submit()
+        await self.submit_condor()
+
+        
+    def modify_timeout(self):
+        printNext = False
```
Comment (mnfienen) [inline objects/Condor.py:23]: you might consider some of these functions/methods that are only called internally to be less public in scope by naming like `_modify_timeout()` etc. More stylistic probably than having meaningful performance impacts, but consider it

File: nwtopt.sub (hunk 5 above)
Comment (mnfienen) [reply nwtopt.sub:16]: ok - now it seems this got replaces when you ran. Consider making a template version (like `nwtopt.submit.template` )with the placeholders for <ip> and <port> and reading that with the code which then writes the actual sub file?

File: objects/Condor.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import os
+import fileinput
```
Comment (mnfienen) [inline objects/Condor.py:2]: sick! never seen this `fileinput` package - imma use that! But.....see above about the template idea for the sub file. Might get sticky with tracking real IP addresses

File: objects/DB_Poller.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        except AllTrialsFailed:
+            pass
+        # get the rest of the trials
+        for i in range(len(trials.trials)):
+            self.HParams.HParam2File(trials.trials[i].get('misc').get('vals').to_dict(), 
+                                     os.path.join(self.cwd, f'{self.key}_{self.HParams.filetype}s', f'{self.HParams.filetype}_{i}.{self.HParams.filetype}'))
+
+        # get results of each trial and compile them into a list
+        results = []
+        minLoss = 9999999
```
Comment (mnfienen) [inline objects/DB_Poller.py:48]: consider `np.inf` to be 100% sure you don't have a legit loss function greater than the initialization (I know, you won't......just style)

File: objects/Master.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+# import from directory up
+sys.path.append('..')
+from NWT_SUBMIT.NWTOPT_FILES.objective import objective
+from config.HParams import MF5HParams
+
+class Master(OPTSubprocess):
+    def __init__(self, id, logger, cwd, ip, port, key, random, trials):
+        super().__init__('master', id, logger)
+        self.cwd = cwd
+        try:
```
Comment (mnfienen) [inline objects/Master.py:17]: you can also do like this to avoid the `try/except`

```python
if os.path.exists(somedir): 
    shutil.rmtree(somedir) 
os.mkdir(somedir)

I don't like relying on error trapping for more known behavior

File: run.sh [hunk 10]
```diff
@@ -1,6 +1,6 @@
 #!/bin/sh
 ############################
 # Run Command:
-./mfnwt meras_transient_rjh_19Feb20.nam
+wine MODFLOW-NWT_64.exe CP_modflow.nam
```
Comment (mnfienen) [inline run.sh:4]: see comment above about maybe making template versions of these files to avoid tracking specifics

**Review by mnfienen — CHANGES_REQUESTED** [review]
This is a solid refactor and looks really good overall! I like the class structure and made a couple subtle comments on it

A couple more important general things:
 0) really need docstrings for the classes and methods
 1) still very few comments - a little would go a long way!
 2) the technique of replacing strings in files is slick but you  will want to populate them with general rather than specific values for things like IP, PORT, model exe name, etc. As a result, but have to go oldskool and have a template version that you read in and fully write out the updated version. This will help people debugging and make it more clear (upon just reading the files) what things are hard-coded and what are variable (e.g. I like <ip> and <port> kind of syntax.
  3) I couldn't really speak to the `async` behavior - it looks to make sense but I've never used it before so not sure about whether it works (I assume you tested) or whether there are stylistic or performance issues.
  4) finally, in the "explicit is better than implicit" vein from the Zen of python (try `import this` in a python interpreter if you haven't seen it!) I think you can avoid some of the `try/except` traps and more explicitly test for things like existence of files/folders. I commented on one specific case above.

File: config/HParams.py (hunk 2 above)
Comment (maxwnewcomer) [reply config/HParams.py:45]: I see what you're saying, but this method is called in a loop and creating new objects (or reassigning attributes) seems less efficient than simply passing them in. The concept behind the HParam object was more of a template for hparams and the functions they need, and not an object of specific HParams.

Comment (maxwnewcomer) [reply config/HParams.py:45]: this should've also been clarified in docstrings :(((( lol

File: nwtopt.sub (hunk 5 above)
Comment (maxwnewcomer) [reply nwtopt.sub:16]: Yeah this was a mistake in the commit, I'll revert it back (thought I had .gitignore not tracking the file, but apparently it is??)

File: objects/Condor.py (hunk 7 above)
Comment (maxwnewcomer) [reply objects/Condor.py:2]: yeah it's kind of nice! and yeah the sub file will be changed back.

File: objects/DB_Poller.py (hunk 8 above)
Comment (maxwnewcomer) [reply objects/DB_Poller.py:48]: 100% Agree, make it look pretty :)

File: run.sh (hunk 10 above)
Comment (maxwnewcomer) [reply run.sh:4]: yeah again i thought this was in .gitignore, don't know why it keeps slipping into the commits.

File: objects/Master.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        self.hparams = MF5HParams().hparams
+        self.ip = ip
+        self.port = port
+        self.key = key
+        self.random = random
+        self.mongoTrials = None
+        self.trials = trials
+
+        try:
+            os.remove(os.path.join(self.cwd, 'NWT_SUBMIT', 'PROJECT_FILES', 'nwts', 'nwtnum.txt'))
```
Comment (mnfienen) [inline objects/Master.py:38]: again, wouldn't rely on try/except here. You can just test to see if the file is there and, if it is, delete it

File: NWT_SUBMIT/NWTOPT_FILES/objective.py [hunk 12]
```diff
@@ -298,18 +299,18 @@ def objective(inputHp):
...
+                'mass_balance': INF,
                 'sec_elapsed': timelim,
                 'iterations': -1,
                 'NWT Used': pathtonwt,
                 'finish_time': finish_time}
     # if no errors get run results
     sec_elapsed, iterations, mass_balance = getRunResults(cwd, listfile)
-    if mass_balance == 999999:
-        loss = 999999999999
+    if mass_balance == INF:
```
Comment (mnfienen) [inline NWT_SUBMIT/NWTOPT_FILES/objective.py:312]: does this work? you might need to do `if np.isinf(mass_balance)`. please verify

File: conda_requirements.yml [hunk 13]
```diff
@@ -1,56 +1,56 @@
...
-- wheel=0.36.2
-- xz=5.2.5
-- zlib=1.2.11
-- pip
-- pathlib=1.0
-- dill=0.3.4
-- wrapt=1.12.1
-- multiprocess=0.70.12.2
-
+  - _libgcc_mutex=0.1=main
```
Comment (mnfienen) [inline conda_requirements.yml:6]: this level of specificity will not be platform agnostic. Might want to simplify it a bit (e.g. just the higher level packages and let conda grab the deps)

File: conda_requirements.yml [hunk 14]
```diff
@@ -1,56 +1,56 @@
...
+  - wheel=0.36.2=pyhd3eb1b0_0
+  - xz=5.2.5=h7b6447c_0
+  - zlib=1.2.11=h7b6447c_3
+  - pip:
+    - dill==0.3.4
+    - multiprocess==0.70.12.2
+    - pathlib==1.0.1
+    - wrapt==1.12.1
+    - wrapt-timeout-decorator==1.3.1
+prefix: /home/mnewcomer/anaconda3/envs/NWTOPT
```
Comment (mnfienen) [inline conda_requirements.yml:56]: also if someone tries to build this env I think they will have a problem because of this path. related to specificity noted above

File: nwtopt.sh [hunk 15]
```diff
@@ -19,16 +19,24 @@ source nwtenv/bin/activate
...
 cp run.sh NWT_SUBMIT/PROJECT_FILES/run.sh
+mv NWT_SUBMIT/NWTOPT_FILES/HParams.py NWT_SUBMIT/PROJECT_FILES/HParams.py
 mv NWT_SUBMIT/NWTOPT_FILES/mfnwt NWT_SUBMIT/PROJECT_FILES/mfnwt
 mv NWT_SUBMIT/NWTOPT_FILES/nwtnum.txt NWT_SUBMIT/PROJECT_FILES/nwts/nwtnum.txt
 mv NWT_SUBMIT nwtenv/bin/NWT_SUBMIT
 echo ...Moved Necessary Files.....
 echo
-ls
 dir=$(pwd)
+export PYTHONPATH="${PYTHONPATH}:${dir}/nwtenv/bin/NWT_SUBMIT"
```
Comment (mnfienen) [inline nwtopt.sh:38]: FWIW I try not to rely on the `PYTHONPATH` var and, instead, smack my conda pack env at the front of the system path var `$PATH` like `export PATH=$PWD/nwtenv/bin:$PATH` but maybe you are relying on it for more than running the python binary?

**Review by mnfienen — COMMENTED** [review]
Looking good! #docstrings

There's still always ideas, but overall looks like it will work!

File: nwtopt.sh (hunk 15 above)
Comment (maxwnewcomer) [reply nwtopt.sh:38]: Yeah that was remnant code from trying to get the hyperparams to transfer to workers. I believe its the solution, but for this merge it is not necessary

File: NWT_SUBMIT/NWTOPT_FILES/objective.py (hunk 12 above)
Comment (maxwnewcomer) [reply NWT_SUBMIT/NWTOPT_FILES/objective.py:312]: yep this works!

File: nwtopt.sub (hunk 5 above)
Comment (maxwnewcomer) [reply nwtopt.sub]: I did, you must be still seeing the outdated file 
image. If you look at actual final commit it is changed back

File: objects/Master.py (hunk 11 above)
Comment (maxwnewcomer) [reply objects/Master.py:38]: I think you are looking at old commit still, new one for me has path.exists all over.

File: conda_requirements.yml (hunk 14 above)
Comment (maxwnewcomer) [reply conda_requirements.yml]: Sounds good, I'll fix

File: nwtopt.sh (hunk 15 above)
Comment (maxwnewcomer) [reply nwtopt.sh]: Fixed in most recent commit

File: conda_requirements.yml (hunk 14 above)
Comment (maxwnewcomer) [reply conda_requirements.yml]: Fixed in most recent commit

**Review by mnfienen — APPROVED** [review]
I think everything is covered in here
